cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/DistributedFlowletInstanceUpdater.java

```diff
@@ -56,16 +63,34 @@
     this.txExecutorFactory = txExecutorFactory;
   }
 
-  void update(String flowletId, int newInstanceCount, int oldInstanceCount) throws Exception {
-    waitForInstances(flowletId, oldInstanceCount);
-    twillController.sendCommand(flowletId, ProgramCommands.SUSPEND).get();
+  void update(String flowletId, int newInstanceCount, FlowSpecification flowSpec) throws Exception {
+    // Find all flowlets that are source of the given flowletId.
+    Set<String> flowlets = getUpstreamFlowlets(flowSpec, flowletId, Sets.<String>newHashSet());
+    flowlets.add(flowletId);
 
-    FlowUtils.reconfigure(consumerQueues.get(flowletId),
-                          FlowUtils.generateConsumerGroupId(program, flowletId), newInstanceCount,
-                          streamAdmin, queueAdmin, txExecutorFactory);
-
-    twillController.changeInstances(flowletId, newInstanceCount).get();
-    twillController.sendCommand(flowletId, ProgramCommands.RESUME).get();
+    try {
+      // Suspend all upstream flowlets and the flowlet that is going to change instances
+      for (String id : flowlets) {
+        waitForInstances(id, getInstances(flowSpec, id));
+        // Need to suspend one by one due to a bug in Twill (TWILL-123)
+        twillController.sendCommand(id, ProgramCommands.SUSPEND).get();
+      }
+      FlowUtils.reconfigure(consumerQueues.get(flowletId),
+                            FlowUtils.generateConsumerGroupId(program, flowletId), newInstanceCount,
+                            streamAdmin, queueAdmin, txExecutorFactory);
+      twillController.changeInstances(flowletId, newInstanceCount).get();
+    } finally {
+      for (String id : flowlets) {
+        // Need to resume one by one due to TWILL-123.
+        // Also ignore any error as resuming a non-suspended flowlet is ok,
+        // except getting from exception log from the container.
+        try {
+          twillController.sendCommand(id, ProgramCommands.RESUME).get();
+        } catch (Exception e) {
+          LOG.warn("Failed to resume flowlet {}", id, e);
```

Comment:
maybe you should error out here? Because it means one flowlet is not running. So the flow is dysfunctional.